include/aws/common/encoding.h

```diff
@@ -138,6 +138,14 @@ enum aws_text_encoding {
     AWS_TEXT_ASCII,
 };
 
+/**
+ * Callbacks to add another parse to customized validation for UTF-8 code point.
+ * @param codepoint the codepoint for validation. We assume that the codepoint
+ *                  would already be respecting RFC-3629.
+ * @return true if the codepoint is valid, otherwise false
+ */
+typedef bool(aws_utf8_customized_validator_callback_fn)(uint32_t codepoint);
```

Comment:
```suggestion
/**
 * Callback invoked for each Unicode codepoint in a UTF-8 string.
 * Use this callback to store codepoints as they're decoded.
 * or to perform additional validation. RFC-3629 is already enforced,
 * which forbids codepoints between U+D800 and U+DFFF,
 * but you may need to forbid codepoints like U+0000.
 *
 * Return AWS_OP_SUCCESS to continue processing the string,
 * Raise an error (i.e. AWS_ERROR_INVALID_UTF8) to stop processing
 * the string and report failure.
 */
typedef int(aws_utf8_on_codepoint_fn)(uint32_t codepoint, void *user_data);
```

1) All callbacks must take a `void *`. If you leave this out, the callback can't meet a lot of use cases (i.e. calling back into another language)
2) The `void *` also opens this up to being used as a "decoder" if anyone ever needed to store the codepoints as their decoded.
3) let's return `int` instead of `bool` so someone can report their own special errors if they need to (like, if they're decoding a string, and it passes some max length, they'd want to raise a specific error about that, not just INVALID_UTF8)